### PR TITLE
Rule S3903: Types should be defined in named namespaces

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            if (declaration.Parent.Parent is TypeBlockSyntax || // Inner type
+                declaration.Parent.Parent is NamespaceBlockSyntax) // Outer type within a namespace
+            {
+                return false;
+            }
+
+            // If we are an outer type without a declared namespace,
+            // make sure there is no Root Namespace set in the project
+            var typeSymbol = (INamedTypeSymbol)semanticModel.GetDeclaredSymbol(declaration);
+            return typeSymbol != null && typeSymbol.ContainingNamespace.IsGlobalNamespace;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs]: VB.NET has this "Root Namespace" setting that allows you not to declare types in namespaces and they will automatically go into that namespace. It is optional, though, so we need to check the symbol.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                GetAnalysisAction(rule),
+                SyntaxKind.ClassStatement,
+                SyntaxKind.StructureStatement);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs]: The `*Statement` is the first line of a `Class/StructBlock`. Registering for the inner node spares us one null check...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DeclareTypesInNamespaces.cs [hunk 3]
```diff
@@ -30,32 +30,25 @@ namespace SonarAnalyzer.Rules.CSharp
...
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(rule);
+
+        protected override SyntaxToken GetTypeIdentifier(SyntaxNode declaration) =>
+            ((TypeDeclarationSyntax)declaration).Identifier;
 
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, declaration.Identifier.GetLocation(),
-                        declaration.Identifier.ValueText));
-                },
+        protected override bool IsOuterTypeWithoutNamespace(SyntaxNode declaration, SemanticModel semanticModel) =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DeclareTypesInNamespaces.cs]: There is something I don't understand, methods indicate that type is NOT in a namespace and yet we return `true` is node parent is `NamespaceDeclarationSyntax`. What am I missing?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        protected override bool IsOuterTypeWithoutNamespace(SyntaxNode declaration, SemanticModel semanticModel)
+        {
+            if (declaration.Parent.Parent is TypeBlockSyntax || // Inner type
+                declaration.Parent.Parent is NamespaceBlockSyntax) // Outer type within a namespace
+            {
+                return false;
+            }
+
+            // If we are an outer type without a declared namespace,
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs]: Missing words on your sentence.

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs]: [cosmetic][minor] I don't understand why you line break so soon compared to your method signature length.

Comment (valhristov) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DeclareTypesInNamespaces.cs]: To have similar length of the two lines. There are no missing words :)

**Pull request merged**